src/server/exchangeToken.ts

```diff
@@ -0,0 +1,68 @@
+import Axios from 'axios'
+import Router from 'koa-router'
+import { ParameterizedContext } from 'koa'
+import { createSession } from 'shared/sessionStorage'
+import { ServerCookieStorage } from 'utils/storage/ServerCookieStorage'
+import { WithLoggerState } from './middleware/enhancers'
+
+enum SearchParams {
+  AuthorizationCode = 'authorizationCode',
+  Next = 'next',
+}
+
+export const exchangeTokenMiddleware: Router.IMiddleware<
+  WithLoggerState,
+  any
+> = async (ctx: ParameterizedContext) => {
+  const nextPath = ctx.URL.searchParams.get(SearchParams.Next)
+  if (!nextPath) {
+    return ctx.throw(400, `Search param ${SearchParams.Next} is required`)
+  }
+  // Safe way of building redirect URL.  We don't want to redirect to external site
+  const nextUrl = new URL('', ctx.URL)
+  nextUrl.pathname = nextPath
+  nextUrl.searchParams.delete(SearchParams.Next)
+  nextUrl.searchParams.delete(SearchParams.AuthorizationCode)
+
+  const authorizationCode = ctx.URL.searchParams.get(
+    SearchParams.AuthorizationCode,
+  )
+  if (!authorizationCode) {
+    return ctx.throw(
+      400,
+      `Search param ${SearchParams.AuthorizationCode} is required`,
+    )
```

Comment:
What happens when a user ends up here? Just thinking it would be nice to show an error page in the future...